transport/dnsovertcp/packet_listener.go

```diff
@@ -0,0 +1,349 @@
+		return 0, errInvalidDNSMsg
+	}
+
+	conn, n, err := c.sendDNSRequestOverTCP(p, addr)
+	if err != nil {
+		return
+	}
+
+	go c.listenDNSResponseOverTCP(conn, addr)
+	return
+}
+
+func (c *dnsOverTcpConn) LocalAddr() net.Addr {
+	// Lock() for safely reading the `c.conns` list
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.conns.Len() > 0 {
+		return c.conns.Front().Value.(net.Conn).LocalAddr()
+	}
+	return nil
+}
+
+// SetDeadline is equivalent to calling both SetReadDeadline and SetWriteDeadline.
+// A zero value for t means I/O operations will not time out.
+func (c *dnsOverTcpConn) SetDeadline(t time.Time) error {
+	return errors.Join(c.SetReadDeadline(t), c.SetWriteDeadline(t))
+}
+
+// SetReadDeadline sets the deadline for future ReadFrom calls and any currently-blocked ReadFrom call.
+// A zero value for t means ReadFrom will not time out.
+func (c *dnsOverTcpConn) SetReadDeadline(t time.Time) error {
+	// Lock() for reading the `c.conns` list and updating recvDdl
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.alreadyClosed() {
+		return errClosed
+	}
+
+	c.recvDdl.SetDeadline(t)
+	return c.forEachConnIgnoreErrClosed(func(conn net.Conn) error {
+		return conn.SetReadDeadline(t)
+	})
+}
+
+// SetWriteDeadline sets the deadline for future WriteTo calls and any currently-blocked WriteTo call.
+// A zero value for t means WriteTo will not time out.
+func (c *dnsOverTcpConn) SetWriteDeadline(t time.Time) error {
+	// Lock() for reading the `c.conns` list and updating wrDdl
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.alreadyClosed() {
+		return errClosed
+	}
+
+	c.wrDdl.Store(&t)
+	return c.forEachConnIgnoreErrClosed(func(conn net.Conn) error {
+		return conn.SetWriteDeadline(t)
+	})
+}
+
+func (c *dnsOverTcpConn) alreadyClosed() bool {
+	select {
+	case <-c.done:
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *dnsOverTcpConn) forEachConnIgnoreErrClosed(w func(net.Conn) error) error {
+	errs := make([]error, 0, c.conns.Len())
+	for e := c.conns.Front(); e != nil; e = e.Next() {
+		err := w(e.Value.(net.Conn))
+		if errors.Is(err, net.ErrClosed) {
+			err = nil
+		}
+		errs = append(errs, err)
+	}
+	return errors.Join(errs...)
+}
+
+func (c *dnsOverTcpConn) sendDNSRequestOverTCP(req []byte, resolver net.Addr) (e *list.Element, n int, err error) {
+	ctx, cancel := context.WithDeadline(context.Background(), *c.wrDdl.Load())
+	defer cancel()
+
+	// Create TCP conn and add it to c.conns
+	conn, err := c.sd.Dial(ctx, resolver.String())
+	if err != nil {
+		return
+	}
+	c.mu.Lock()
+	e = c.conns.PushFront(conn)
+	c.mu.Unlock()
+
+	// If there are errors, remove conn from c.conns
+	defer func() {
+		if e != nil && err != nil {
+			conn.Close()
+			c.mu.Lock()
+			c.conns.Remove(e)
+			c.mu.Unlock()
+			e = nil
+		}
+	}()
+
+	if err = conn.SetReadDeadline(c.recvDdl.Deadline()); err != nil {
+		return
+	}
+	if err = conn.SetWriteDeadline(*c.wrDdl.Load()); err != nil {
+		return
+	}
+
+	// DoT: message length prefix
```

Comment:
DoT usually refers to DNS-over-TLS:
```suggestion
	// DNS-over-TCP: message length prefix
```